### ISSUE TITLE
Support Last-Modified header

### ISSUE THREAD

**Issue by osma** [NatLibFi/Skosmos#574]
Skosmos does not currently set the Last-Modified HTTP header on responses. The data comes directly from the triple store and Skosmos is generally unaware of when it has last changed.

However, some vocabularies include dct:modified timestamps, either for the whole vocabulary (set on the skos:ConceptScheme resource) and/or individual concepts. Skosmos could make use of these timestamps to set Last-Modified headers, at least for concept pages (and RDF data for concepts) but possibly also other responses. For a concept page, the Last-Modified timestamp would primarily be based on the concept-level timestamp, but if not set, the concept scheme level timestamp could be used instead.

This would perhaps have to be enabled in a per-vocabulary setting. Not sure if there are any dangers in enabling it for all vocabularies. There would be some additional SPARQL queries, but they should be fast and easily cached.

**Comment by kinow** [comment]
I this one, #574, and #575, might be done together.

I have a branch for the performance work around mapping properties, and thought it would be actually simple to try something on this one using that branch (because I had Docker + cache + config.ttl properly setup, etc).

Here's what I did:

1. Added a vocabulary config entry called "useModifiedDate"... not happy with the value... I guess someone could come up with a better name for this setting? I have the feeling that it might be either confusing or cause trouble in the future if we add more settings?
2. Added a method to `Concept` to return the modified date
3. Added some logic to the `WebController` where it returns concepts, to search for the vocab config setting (1.), and the concept's modified date (2.), then also check if the browser (or any other client, tested curl, and it worked fine) set the `If-Modified-Since` header.
4. Finally, if the header is present, and if the content has not changed, then return an empty 304. But if the header is not present, then (and this is important) set the `Last-Modified` header. When browsers receive the `Last-Modified` header, they send it back as `if-Modified-Since` in the next request, unless the user marked to disable cache or forced the reload of the page...

Screenshots below show the initial request receiving a 200 with the Last-Modified, then a subsequent request/response 304, being correctly displayed by the browser.

![screenshot from 2019-01-03 00-14-04](https://user-images.githubusercontent.com/304786/50590051-1ba03400-0eee-11e9-8037-b20da278bd1e.png)
![screenshot from 2019-01-03 00-16-52](https://user-images.githubusercontent.com/304786/50590050-1ba03400-0eee-11e9-808a-c13563e9b87a.png)
![screenshot from 2019-01-03 00-17-08](https://user-images.githubusercontent.com/304786/50590048-1ba03400-0eee-11e9-82b3-573b4236b912.png)
![screenshot from 2019-01-03 00-17-31](https://user-images.githubusercontent.com/304786/50590047-1b079d80-0eee-11e9-876d-e4ecef046d9e.png)
![screenshot from 2019-01-03 00-17-54](https://user-images.githubusercontent.com/304786/50590045-1b079d80-0eee-11e9-8da2-2130587d8223.png)

Tested with curl, Firefox, and Chrome, on Ubuntu LTS.

@osma, does this sound like going on the right direction? I stopped here, leaving the changes stashed in my local branch because it's a mess and it would still require a) cleaning the code, b) adding some comments perhaps, c) writing tests! Will continue if someone confirms it sounds like a good plan, or if there are other things that I am missing (e.g. still have to use the conceptscheme modified date too).

Cheers
Bruno

Cheers

**Comment by kinow** [comment]
Hi Osma!

>1. A vocabulary that has dct:modified timestamps for all Concepts. E.g. YSO Places that you've used in your example.

Right. That's done right now.

>2. A vocabulary that has dct:modified timestamps for some Concepts, but not all. E.g. YSO. (Don't ask why...) Then it would make sense to fall back on the timestamp of the ConceptScheme.
> 3. A vocabulary that doesn't have timestamps for Concepts, but has a timestamp for the ConceptScheme. Like above, we should fall back on that. E.g. Lapponica and SLM from Finto are like this.

Right. Sounds like a good plan. Simply try dct:modified. If not found, then try the ConceptScheme's timestamp. If not found, just ignore and return the normal Concept without the Last-Modified header.

>A vocabulary that doesn't have any timestamps. Then I guess we can't produce a Last-Modified header. E.g. most of the toy vocabularies used for unit tests don't have timestamps at all. Also Kassu and IPTC vocabularies in Finto are like this.

Agreed.

>Unit tests should of course cover all above cases.

Sure thing, we can probably achieve near 100% coverage for the new code. Only issue being that testing request/response with&without headers is a bit troublesome to build. But once one example is done, it should be straightforward to test the other cases.

>It's also possible to encounter malformed timestamps (e.g. 1980-00-00). I believe e.g. YSA has some of these, artifacts of conversion from MARC records. It would probably make sense to ignore such timestamps, as if they didn't exist at all.

Hummm... good point. Or timestamps in other formats (10/10/2012...). I reckon these should be simply ignored.

Sounds like we have a plan! So next steps I believe will be to cover the cases above, write the unit tests, and create a pull request. WDYT?

Cheers
Bruno

**Comment by osma** [comment]
I've thought about this some more and I think there may be a problem with the current approach, which is to take the timestamp from the concept (or concept scheme) data and simply use that. The problem is that the Last-Modified header really applies for the whole HTTP response, which in practice is an HTML document. The generated response depends on not only the concept data, but also e.g. the version of Skosmos and its configuration. A new version of Skosmos may well generate a different HTML document even if the concept data stays unchanged.

So I wonder if the timestamp generation should also consider other timestamps, especially the current version of Skosmos. One way to check that is to execute the command `git log -1 --date=iso --pretty=format:%cd`, which outputs the timestamp of the last Git commit. Another possible timestamp to consider is the last modified date of the `config.ttl` file (note that if `config.ttl` is a symlink, it should be followed and check the timestamp of the target file instead of the symlink).

The actual returned Last-Modified timestamp could be the most recent timestamp out of the three sources of timestamp information: 1. concept or concept scheme timestamp (as currently), 2. timestamp of last Git commit and 3. last modified date of `config.ttl`.

Of course, this means that the timestamps will tend to be fairly recent, but I think that the Skosmos code and configuration should still usually be several days or weeks old in a production system such as finto.fi. I just checked, and finto.fi is running Skosmos code dated `2019-01-11` and the `config.ttl` timestamp is `2019-01-25` while the previous change was on `2019-01-14`.

**Comment by kinow** [comment]
Hi @osma 

I don't have much experience managing a Skosmos installation, but what you said sounds good to me. And I think it's doable. I tried to keep this pull request's code coverage as close to 100% patch coverage as possible, so will just need to a) implement the code for the 2 new timestamps (git & config.ttl last modified), b) test locally, and c) write some unit tests.

Might take a few more days, but hopefully not really long :slightly_smiling_face: 

Cheers
Bruno

**Comment by kinow** [comment]
Hi @osma,

Had some spare time today, so had a go at implementing these new modified dates.

https://github.com/kinow/Skosmos/compare/ac5bab2ef594d5959fb336da862a9495cdde6c6c...1e233fec178479dc0c6d3ee050918182ecd8440d#diff-2d3c487a3a39abb75e632bf45ef35bf0R262

Basically, used the command you provided, and also `filemtime` from PHP. Spent most of the time actually designing how the methods should be organized, so that testing was a bit easier to write and to understand later too.

Managed to cover the whole new test with simple unit tests through some mockery. The unit tests passed locally, so I rebased it, and also enabled Travis-CI in my branch to check what's the build status ([passed](https://travis-ci.org/kinow/Skosmos/builds/497725358)).

I will have time to test it thoroughly throughout this week or maybe next weekend. But if you have spare time to review Skosmos issues, maybe you could look into this one (or that other one about concept mapping properties).

Cheers
Bruno

**Comment by osma** [comment]
Hi @kinow ,

Finally found some time to look at this. The link in the above comment didn't seem to work but I grabbed the code in your branch [support-last-modified-header](https://github.com/kinow/Skosmos/tree/support-last-modified-header) and tried to run it locally.

I didn't get very far though: I'm getting this error:

    Fatal error: Uncaught Error: Cannot access private property Model::$globalConfig in /var/www/html/Skosmos/controller/WebController.php on line 315

Also I noticed this in WebController.php:

    $cache->store($cacheKey, $commitDate, Model::URI_FETCH_TTL);

so it's using the URI_FETCH_TTL for the TTL value, which is set to 1 day (86400 seconds) in Model.php. I don't think that's ideal here - the `git log` command shouldn't be that expensive (less than 5 ms on my system) so it can be run, say, once per minute (or 10 minutes). After all the code may be updated any time.

Can you please make a PR from this branch, it would be easier to review that way?

**Comment by kinow** [comment]
PR on its way @osma! Thanks!!

**Comment by kinow** [comment]
I think this one can be closed now :+1:

**Comment by osma** [comment]
Well yes, except Last-Modified and 304 are only implemented for concept pages, not e.g. RDF responses that are handled by `RestController` as was specified in the OP. But maybe it's best to close this one (and #575) and open a new issue specifically for RDF responses.

**Comment by osma** [comment]
Opened #864 for RDF responses. Closing this one.
